lib/executor/variable_arrival_rate.go

```diff
@@ -386,49 +300,59 @@ func (varr VariableArrivalRate) Run(ctx context.Context, out chan<- stats.Sample
 
 		return math.Min(1, float64(spent)/float64(duration)), right
 	}
+
 	varr.progress.Modify(pb.WithProgress(progresFn))
 	go trackProgress(ctx, maxDurationCtx, regDurationCtx, varr, progresFn)
 
 	regDurationDone := regDurationCtx.Done()
-	runIterationBasic := getIterationRunner(varr.executionState, varr.logger, out)
-	runIteration := func(vu lib.VU) {
-		runIterationBasic(maxDurationCtx, vu)
-		vus <- vu
-	}
+	runIteration := getIterationRunner(varr.executionState, varr.logger, out)
 
 	remainingUnplannedVUs := maxVUs - preAllocatedVUs
-	rateChangesStream := varr.streamRateChanges(maxDurationCtx, startTime)
 
-	for {
+	var timer = time.NewTimer(time.Hour)
+	var start = time.Now()
+	var ch = make(chan time.Duration, 10) // buffer 10 iteration times ahead
+	var prevTime time.Duration
+	go varr.config.cal(varr.executionState.ExecutionTuple, ch)
+	for nextTime := range ch {
 		select {
-		case rateChange := <-rateChangesStream:
-			newPeriod := rateChange.tickerPeriod
-			ticker.Stop()
-			if newPeriod.Valid {
-				ticker = time.NewTicker(time.Duration(newPeriod.Duration))
-			}
-			atomic.StoreInt64(tickerPeriod, int64(newPeriod.Duration))
-		case <-ticker.C:
-			select {
-			case vu := <-vus:
-				// ideally, we get the VU from the buffer without any issues
-				go runIteration(vu)
-			default:
-				if remainingUnplannedVUs == 0 {
-					//TODO: emit an error metric?
-					varr.logger.Warningf("Insufficient VUs, reached %d active VUs and cannot allocate more", maxVUs)
-					break
-				}
-				vu, err := varr.executionState.GetUnplannedVU(maxDurationCtx, varr.logger)
-				if err != nil {
-					return err
-				}
-				remainingUnplannedVUs--
-				atomic.AddUint64(&initialisedVUs, 1)
-				go runIteration(vu)
-			}
 		case <-regDurationDone:
 			return nil
+		default:
+		}
+		atomic.StoreInt64(&tickerPeriod, int64(nextTime-prevTime))
+		prevTime = nextTime
+		b := time.Until(start.Add(nextTime))
+		if b > 0 { // TODO: have a minimal ?
+			timer.Reset(b)
+			select {
+			case <-timer.C:
+			case <-regDurationDone:
+				return nil
+			}
+		}
+
+		var vu lib.VU
+		select {
+		case vu = <-vus:
+			// ideally, we get the VU from the buffer without any issues
+		default:
+			if remainingUnplannedVUs == 0 {
+				//TODO: emit an error metric?
+				varr.logger.Warningf("Insufficient VUs, reached %d active VUs and cannot allocate more", maxVUs)
+				continue
+			}
+			vu, err = varr.executionState.GetUnplannedVU(maxDurationCtx, varr.logger)
```

Comment:
I realized this (and my original) implementation of initializing unplanned VUs in the middle of the iteration loop is very wrong... This synchronous operation can significantly derail the actual iteration pace... Another reason to implement the first part of https://github.com/loadimpact/k6/issues/1386, or just allocation of the VU in a goroutine. But then again, if a few iterations in a row lack a VU to run on, we probably shouldn't allocate more than 1 VU concurrently...